### ISSUE TITLE
New FBConnect Interface

### ISSUE THREAD

**Issue by dagerydoo** [ideashower/ShareKit#233]
Hello,

I recently altered SHKFacebook to use the new FBConnect interface. In short, it supports the new style of authentication where the user gets swapped to either the native facebook app (if installed) or m.facebook.com. 
### Altered files
- SHKConfig.h (replacing the existing facebook API settings with the lone SHKFacebookAppID). 
- SHKFacebook.h
- SHKFacebook.m
- The entire FBConnect directory
### Removed files
- SHKFBStreamDialog.h (the new Facebook object has a built-in dialog method, which makes subclassing to set the default message problematic, though setting the message is likely possible)
- SHKFBStreamDialog.m 

Also, (unfortunately) the app delegate needs to handle incoming URL's (which is fairly trivial, but it adds an installation step) so that it can propagate the authentication. In my app, I delegated directly to a new class method of SHKFacebook, but I could also imagine a general SHK solution being implemented, as this might be a new model going forward (?).

The accessToken and expirationDate are serialized to NSUserDefaults, so the user does not have to repeatedly authorize the FB app  (whether this is safe is worth considering, though it seems only slightly less safe than storing confidential values in the executable). Finally, the SHKItem is also serialized, so that the user can immediately continue posting the item following the external authentication. I imagine these changes could also lend to the efforts of offline sharing for FB. 

Is there any interest in these changes? I made the effort after running into problems using the original code.

Cheers!

Andrew

**Comment by wimdenood** [comment]
Yes, I'm very interested!! It's not working for me right now. Can you share the code?? My mail address is: iphone@greatadventure.nl

**Comment by lenkaencore** [comment]
hello, could you please share the code? My email is lenka@engage-encore.com. Thanks.

**Comment by nmilosevic** [comment]
I'm interested too. My email is zastava128@gmail.com

**Comment by tobytaro** [comment]
please share the code, thanks.
yzx2002@gmail.com

**Comment by dagerydoo** [comment]
I am cleaning up the code right now, and adding support for picture and text messages (my app only required links). Shouldn't be long. I'll post as a patch...

**Comment by lenkaencore** [comment]
RE dagerydoo: I read that the whole Share feature has been deprecated and that now Like button with user customised message should be used. Is this how you implement it? Or did you get around it some other way? I am just interested because we are considering options with my colleagues right now.

**Comment by dagerydoo** [comment]
I have a patch coming. Here is the fork:

https://github.com/dagerydoo/ShareKit/tree/NewFBConnect

**Comment by dagerydoo** [comment]
Here is the compressed patch file:

http://dl.dropbox.com/u/7735151/ShareKit.FBConnect.diff.zip

**Comment by wimdenood** [comment]
Hi  there! It works for me. Just as described in the Readme. However one problem: If I try to post an image this way to facebook, I get the safari with login and after that it returns to my app. But then the image is NOT posted. When I after that post it again, I don't have to login again (previous login was still valid) and then the image IS posted. 

Any idea how to fix this??

Thanks a lot for the update!!

**Comment by lenkaencore** [comment]
Hello, thanks for sharing this. The project works for me on its own but if I try to implement it in a new iPad application, the facebook interface doesn't get open in the modal screen and the sharing doesn't work after opening the 'You already allowed this app' screen. I added the new URL functions to AppDelegate and added URL types and FacebookURLTypeIndex properties into the info.plist file. Is there anything else I need to do?

**Comment by lenkaencore** [comment]
Ok your solution seems to work on simulator < 4, e..g iPad 3.2. but not on 4.2. Would you be able to fix this perhaps?

**Comment by dagerydoo** [comment]
Hmm... I haven't run into any issues. Is your app multitasking?

**Comment by dagerydoo** [comment]
I'll look into the image posting bug.

**Comment by lenkaencore** [comment]
Yes it is. The problem is that multitasking apps open safari which is fine on the first authentication but not fine on any others (it tries to call an non-existing URL). 

I am currently implementing a version where the Facebook share is handled by the application delegate rather than by ShareKit, similarly how it is shown here: http://developers.facebook.com/docs/guides/mobile/ 
I think the main difference is that the Facebook object is not static but an instance variable of the AppDelegate (could be anywhere though I guess).

**Comment by dagerydoo** [comment]
Hmm... Perhaps I erred by making the Facebook object static. I'll give it another shot.

**Comment by lenkaencore** [comment]
If you could please that would be great, thanks. Unfortunately I have a deadline today so cannot improve on this anymore. But it would be great if I could improve my app later.

**Comment by dagerydoo** [comment]
As for the image posting bug, I quickly implemented a fix within the SHKItem dictionaryRepresentation method, however I do not like the solution, as the images left in the cache are only removed when an accompanying call to itemFromDictionary: is made.

While I like this general solution (for one thing, it guarantees the dictionary truly represents the same item — currently, you lose the image) , and its ease of use, I have decided to augment the dictionary only within the SHKFacebook code. My heart breaks a little doing this, but it is the correct decision.

**Comment by dagerydoo** [comment]
The master branch has been updated to support image posting during the initial authorization:

https://github.com/dagerydoo/ShareKit

**Comment by Vicki-zz** [comment]
The update works great for me. Does anyone have a link to where FB officially says the interface changed? Did it change recently, or has FB just stopped supporting some deprecated features that we were using?

**Comment by dagerydoo** [comment]
@Vicki — I don't think they have officially changed to this new mechanism exclusively. In fact, the official github project for FBConnect indicates that the system will fall back to the old style for non-multitasking apps (which didn't seem to work, actually), under Single Sign-On:

https://github.com/facebook/facebook-ios-sdk

It seems like last week's outage with the old system might have been a glitch, though this interface what active the whole time (as far as I could tell). It is unclear when or even if the old system is going to be deprecated, but my gut tells me it will be sooner rather than later (I can't find a link, but there was a headline that I shot past on Google News this weekend about how apps could steal user credentials, and I would imagine FB and others will push the new interface harder to avoid sullying their reputations).
